# Машинное обучение. ВМК МГУ

# Практическое задание 2: Numpy + pandas + matplotlib

## Уровень: **Исследовательский (Research)**

# О формате сдачи

🔷 При решении ноутбука используйте данный шаблон

```
✅ Можно добавлять новые ячейки любых типов
❌ Не нужно удалять текстовые ячейки c разметкой частей ноутбука и формулировками заданий
```

🔷 При оценивании задач учитывается код

```
✅ Задания, в которых необходим код, обычно помечаются фразами "Your code here"/"Ваш код" и аналогичными
❌ Ответы на вопросы без сопутствующего кода оцениваются в 0 баллов
❌ Наличе работоспособного кода в ноутбуке, если на сказано иного, обязательно
```

🔷 При оценивании задач учитываются выводы

```
✅ Задания, в которых необходимы выводы, обычно помечаются фразами Вывод"/"Ответ на вопрос"/"Ваш текст" и аналогичными
✅ Обычно выводы подразумевают под собой текстовый ответ (можно писать markdown, latex).
✅ Сопутствующие изображения, графики, таблички - приветствуются!
❌ При отсутствии выводов задание не засчитается на полный балл
```

- Многие из заданий можно выполнить несколькими способами.
Не существуют единственно верного, но попробуйте максимально
задействовать арсенал pandas и ориентируйтесь на простоту и понятность
вашего кода. Не забывайте, что можно гуглить и что-то искать на stackoverflow, например.

# Подготовка рабочей среды

Сначала установим нужные нам версии библиотек. Мы гарантируем, что в данных версиях задание будет корректно отрабатывать.

После установки нужных версий, возможно, нужно перезагрузить среду (runtime), но скорее всего вам это не понадобится

На скачивание файла и установку понадобится не более 5 минут.

**Важно!**

Устанавливать нужные версии нужно каждый раз, когда создается новый
рантайм. Например, если вы 2 часа подряд делаете это задание, то
подготовить библиотеки достаточно 1 раз. Но если вы, например, начали в
понедельник, затем закрыли/выключили ноутбук, то при продолжении в
среду, вам нужно будет запустить рантайм заново и следовательно заново
установить библиотеки.

**Важно!**
Если вы предпочитаете делать практические задания на своем личном ноутбуке, то проверьте, что вы установили рабочее окружение в соответствии с гайдом .pdf)

In [ ]:
# !!! Данный блок будет работать только в Google-Colab !!!
! gdown 10k8Hwn9kpK9SpK4IEj4-EaWQZqgYT5-Q
! pip install -r /content/requirements_2024_25_for_colab_small.txt

Downloading...
From: https://drive.google.com/uc?id=10k8Hwn9kpK9SpK4IEj4-EaWQZqgYT5-Q
To: /content/requirements_2024_25_for_colab_small.txt
100% 375/375 [00:00<00:00, 1.33MB/s]
  Using cached jedi-0.19.1-py2.py3-none-any.whl.metadata (22 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 9.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 516.3/516.3 kB 31.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 85.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 78.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.5/41.5 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.5/15.5 MB 81.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.7/14.7 MB 82.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 97.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.8/294.8 kB 23.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.

Проверим версию библиотеки:

In [ ]:
import catboost
assert(catboost.__version__ == '1.2.7')

Теперь можно приступать к выполнению задания! :)

# О задании

Задание состоит из трех частей, посвященных работе с табличными данными с помощью библиотеки pandas , визуализации с помощью библиотек matplotlib , seaborn , plotly и векторным вычислениям с помощью библиотеки numpy . Вам предлагается выполнить несколько заданий.

# Часть 1. Pandas (3,5 балла), кросс-проверка

В этой части вы отработаете навыки работы с библиотекой **pandas** .

**Подсказка** Если вы еще не прошли выданный воркбук по pandas , то советуем в начале обратиться к нему. Также рекомендуем сначала решить базовую часть, прежде чем приступать к этой.

**Внимание!** Некоторые задачи требует дополнительного изучения документации, не стесняйтесь искать нужные методы в интернете

Далее потребуется выполнить несколько небольших заданий, покрывающих
базовый функционал библиотеки, который довольно часто пригождается на
первичных этапах анализа данных перед обучением ML-моделей.

In [ ]:
import pandas as pd

In [ ]:
!gdown 1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6

Downloading...
From: https://drive.google.com/uc?id=1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6
To: /content/youtube_trending_mediterranean.csv
100% 6.22M/6.22M [00:00<00:00, 203MB/s]


### **Задание 1 [0.5 баллa]**

Загрузите данные и посмотрите на колонку publishedAt

- **(0.1 балла)** Какого она типа?
- **(0.2 балла)** Измените тип колонки на datetime . Покажите, что после изменения колонка действительно имеет тип datetime .
- **(0.2 балла)** В какой месяц было добавлено больше всего видеороликов?

In [ ]:
df = pd.read_csv("youtube_trending_mediterranean.csv")
df.dtypes["publishedAt"]
df["publishedAt"]= pd.to_datetime(df["publishedAt"])
df['month'] = df['publishedAt'].dt.month
df_new = df.groupby('month').size()
df_new

,0
month,
6,1068
7,2153


**Тип publishedAt: object**
**, больше всего видео было добавлено в июле**

In [ ]:
df.drop('month', axis=1)

,country,video_id,title,publishedAt,channelId,channelTitle,categoryId,trending_date,tags,view_count,comment_count,thumbnail_link,comments_disabled,ratings_disabled,description
0,AL,hxWSp-8jJuo,DON XHONI - KATILE,2022-07-07 22:00:12+00:00,UCZNJXsf2I4y54M_UmMiC6Vw,DON XHONI,24,22.11.07,[none],5130167,3641,https://i.ytimg.com/vi/hxWSp-8jJuo/default.jpg,False,True,Stream or Download now ''KATILE'' 👇 https:/...
1,AL,MnrC9mk2H1o,Dhurata Dora x Elvana Gjata - GAJDE (Official ...,2022-06-30 22:00:11+00:00,UCtzhq8TIfFB9vRAHVqLL9PA,Dhurata Dora,10,22.11.07,[none],8886648,4414,https://i.ytimg.com/vi/MnrC9mk2H1o/default.jpg,False,True,The official video for GAJDE.Stream/Download G...
2,AL,5Ql-G0dTqC0,KIDA x BUTRINT IMERI - AM/PM,2022-07-02 22:00:11+00:00,UCAHVxL9FG3Sp8ltQtLEy_oA,KIDA,10,22.11.07,Butrint Imeri|Kida|Dale|Butrint imeri kida|But...,7230818,7518,https://i.ytimg.com/vi/5Ql-G0dTqC0/default.jpg,False,True,Stream or Download „AM/PM“ here 🎧 https...
3,AL,oUdY0qSavvY,Dafina Zeqiri - Malli,2022-07-07 22:00:12+00:00,UCrv0PJoutcRhHkSMJYoXAIQ,Dafina Zeqiri,10,22.11.07,Dafina|Zeqiri|Dafine Moj|Pa ty|Lule Lule|Duro|...,1406078,827,https://i.ytimg.com/vi/oUdY0qSavvY/default.jpg,False,True,Bzzz Entertainment - https://lnk.to/BzzzStream...
4,AL,I4ZSvYLnOcQ,Tayna x Azet - Tequila,2022-06-30 18:00:26+00:00,UCJBZ-GcWSswmJ0xc9WAKDqw,Friends Entertainment,10,22.11.07,tayna|Friends|Entertainment|rinor hykolli|hip ...,6397676,6063,https://i.ytimg.com/vi/I4ZSvYLnOcQ/default.jpg,False,True,Stream „Tequila“:https://taynaxazet.lnk.to...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3216,NaN,eB-A1Y0Dswc,Thank you for watching,2022-07-05 21:18:33+00:00,UCS5tt2z_DFvG7-39J3aE-bQ,Life of Boris,22,22.11.07,lifeofboris|life of boris|english subtitles|ye...,864216,14431,https://i.ytimg.com/vi/eB-A1Y0Dswc/default.jpg,False,True,New channel: https://www.youtube.com/channel/U...
3217,NaN,KJGTtzccOHU,Pokemon Emerald but every battle is extremely ...,2022-07-05 01:55:10+00:00,UC0VVYtw21rg2cokUystu2Dw,SmallAnt,20,22.11.07,pokemon|smallant|smallant1|challenge|pokemon c...,1311698,1232,https://i.ytimg.com/vi/KJGTtzccOHU/default.jpg,False,True,Pokemon emerald but every battle is a 1v2You s...
3218,NaN,be8mqaTi_DI,Victor Cibrian x Fuerza Regida x Luis R Conriq...,2022-07-04 21:00:36+00:00,UC648rgJOboZlgcDbW00vTSA,Rancho Humilde,10,22.11.07,Rancho Humilde|Legado 7|El De La Guitarra|Fuer...,2256933,597,https://i.ytimg.com/vi/be8mqaTi_DI/default.jpg,False,True,Suscríbete: https://Rancho.lnk.to/RHYoutube\r...
3219,NaN,14aYHc7ybu0,WE STAYED AT A 1 STAR HOTEL FOR 24 HOURS | coc...,2022-07-05 13:40:04+00:00,UCSGoIq_tVESqNYF1Re-zn1Q,HJ Evelyn,24,22.11.07,hjevelyn|Evelyn ha|Emily ha|erica ha|korean si...,922668,2951,https://i.ytimg.com/vi/14aYHc7ybu0/default.jpg,False,True,never again ahaha.... CHECK OUT OUR MERCH: htt...


### **Задание 2 [1 балл]**

Посмотрите на колонку description .

- **(0.3 балла)** Переведите строки в нижний регистр и обрежьте ненужные пробелы справа и слева.
- **(0.5 балла)** Замените url-адреса в описаниях на специальный тэг \<URL\> (прямо как строчку, "\<URL>").
- **(0.2 балла)** Выведите полностью description для видеоролика с video_id = 'Oxexirlmdig' .

In [ ]:
pattern = r'http\S+' # возможно пригодятся регулярные выражения..

In [ ]:
import re
df['description'] = df['description'].str.lower()
df['description'] = df['description'].apply(lambda x: re.sub(r'http\S+', '<URL>', x) if isinstance(x, str) else x)
pd.set_option('display.max_colwidth', None)
df[df['video_id'] == 'Oxexirlmdig']['description'].to_string()

'12    music line & lyrics : ghetto geasy & majkmusic produced : bigdeal records mix & master: bigdeal recordsavailable on:spotify: <URL> <URL> <URL>'


In [ ]:
pd.set_option('display.max_colwidth', 50)

**music line & lyrics : ghetto geasy
& majkmusic produced : bigdeal records mix & master: bigdeal
recordsavailable on:spotify: **

### **Задание 3 [1 балл]**

Какое максимальное число просмотров (колонка view_count ) у видеороликов, выпущенных во Франции или Италии (в колонке country значениe FR или IT) в июле (колонка publishedAt)?

Необходимо придумать решение в одну строку , используя
для выбора строк несколько условий, объединенных с помощью | вместо or,
& вместо and и ~ вместо not, а также используя функцию isin .

In [ ]:
df[(df['country'].isin(['FR', 'IT']))&(df['publishedAt'].dt.month==7)]['view_count'].max()

82864422


**82864422**

### **Задание 4 [0.5 баллa]**

- **(0.2 балла)** Посмотрите на названия каналов из 17 категории (колонка categoryId). Как вы думаете, что это за категория?
- **(0.3 балла)** Предположим, что
была допущена ошибка, и все видеоролики из 17 категории, относятся к 16
категории. Исправьте нужные значения в колонке categoryId, после чего
выведите все уникальные категории.

**Подсказка**
Обратите внимание, что при выборе полей с помощью оператора
индексирования [] возвращается копия выбранных значений. При попытке
присвоить таким образом новые значения для некоторых полей таблицы,
будет выдано предупреждение. Более надежный способ - использовать
операторы loc и iloc . Подробнее про индексацию и выбор строк можно прочитать здесь .

In [ ]:
df[df['categoryId']==17]['channelTitle']
df.loc[df['categoryId']==17, 'categoryId']=16
df['categoryId'].unique()

array([24, 10,  1, 15, 22, 23, 27, 16, 20, 26, 25,  2, 19, 28])


**спортивные каналы**

### **Задание 5 [0.5 баллa]**

Посмотрите на колонку tags .

- **(0.2 балла)** Замените строки со значением '[none]' на пустую строку.
- **(0.3 балла)** Какой самый популярный тэг? Обратите внимание, что у каждого видеоролика может быть более 1 тэга.

**Подсказка**

Чтобы получить правильное значение, не забудьте обработать множество
тэгов - убрать пробельные символы по краям и привести строки в нижний
регистр.

**Подсказка**

Нужный тэг встречается больше 130 раз.

In [ ]:
df.loc[df['tags']=='[none]', 'tags']=""
def edit(text):
  return text.lower().lstrip().rstrip()
df['tags'].value_counts()[df['tags'].apply(edit).mode()]

**611**

# Часть 2. Визуализация данных (0,5 баллa), кросс-проверка

### Оформление графиков

Важно, чтобы все графики выглядели опрятно и соответствовали следующим правилам:

🔷 Максимально поясняйте, что изображено на графике

```
✅ Добавляйте название (plt.title)
✅ Подписывайте оси (plt.xlabel, plt.ylabel)
✅ Подписывайте единицы измерения на осях
✅ Делайте понятные подписи
```

🔷 Следите за визуальной составляющей графика

```
✅ Подписи нигде не должны заезжать друг на друга
✅ Все линии на графиках должны быть чётко видны
❌ Не используйте похожие цвета или цвета, сливающиеся с фоном
✅ Используйте векторный (pdf/svg) формат при возможности
✅ Ограничивайте диапазоны значений там, где это возможно (например, вероятности можно ограничить диапазоном [0,1])
✅ Для улучшения демонстрации различий можно логарифмировать оси x и/или y (plt.xscale("log"), plt.yscale("log"))
❌ Не рекомендуется использовать 3d графики в презентациях: при возможности изобразите необходимое при помощи нескольких 2d графиков
```

🔷 Учитывайте размеры и вместимость графиков

```
✅ Графики должны быть комфортного для показа размера: все, что необходимо показать, должно быть хорошо видно
✅ Иногда можно разбивать логически график на несколько полотен, если на одно все не умещается
```

Больше примеров Вы можете найти в документации .

In [ ]:
import matplotlib
import matplotlib.pyplot as plt

#отображение графиков в ноутбуке
%matplotlib inline

#для четкой прорисовки графиков - осторожно!!! Может съедать много памяти
#закоменьте эту строку, если это случилось
%config InlineBackend.figure_format = 'svg'

Далее вам нужно построить несколько визуализаций табличных данных о видео, с которыми мы работали в 1 части задания.

**Внимание!**

- Проверьте, что при конвертации в html формат все графики остаются видны.
Это нужно для дальнейшей проверки в проверяющей системе. Проверить это
можно непосредственно загрузив задание в систему и посмотреть на свою
отправленную посылку (как это сделать - можно почитать в туториале о
проверяющей системе)
- При решении заданий вы можете воспользоваться любыми средствами для визуализации.
- Все графики должны выглядеть опрятно и соответствовать правилам, перечисленным в начале Части 2. В случае несоответствия хотя бы 3 пунктам оценка может быть снижена

TODO

### **Задание 1 [0,5 балл]**

Изобразите на одном полотне несколько графиков вида box_plot , отображающие: для каждой страны ( country ) box_plot распределения значения просмотров ( view_count ). Причем нужно это сделать только по тем роликам, по которым комментарии запрещены ( comments_disabled )

Ответьте на вопрос, опираясь на график: в каких двух странах в среднем наибольшее число просмотров?

In [ ]:
filtered=df[df['comments_disabled']==True]
plt.figure(figsize=(10, 10))
filtered.boxplot(column='view_count', by='country')
plt.suptitle('')
plt.title('Распр. кол. просмотров видео с выкл. комм. по странам')
plt.yscale("log")
plt.xlabel('Страна')
plt.ylabel('Количество просмотров')

Text(0, 0.5, 'Количество просмотров')


<Figure size 1000x1000 with 0 Axes>


?xml version="1.0" encoding="utf-8" standalone="no"?
 
 
 
 
 
 2024-10-11T16:25:09.298885 
 image/svg+xml 
 
 
 Matplotlib v3.7.1, https://matplotlib.org/ 
 
 
 
 
 
 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- AL -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- BA -->
 
 
<path d="M 1259 2228 
L 1259 519 
L 2272 519 
Q 2781 519 3026 730 
Q 3272 941 3272 1375 
Q 3272 1813 3026 2020 
Q 2781 2228 2272 2228 
L 1259 2228 
z
M 1259 4147 
L 1259 2741 
L 2194 2741 
Q 2656 2741 2882 2914 
Q 3109 3088 3109 3444 
Q 3109 3797 2882 3972 
Q 2656 4147 2194 4147 
L 1259 4147 
z
M 628 4666 
L 2241 4666 
Q 2963 4666 3353 4366 
Q 3744 4066 3744 3513 
Q 3744 3084 3544 2831 
Q 3344 2578 2956 2516 
Q 3422 2416 3680 2098 
Q 3938 1781 3938 1306 
Q 3938 681 3513 340 
Q 3088 0 2303 0 
L 628 0 
L 628 4666 
z
" id="DejaVuSans-42" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- DZ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- EG -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- ES -->
 
 
<path d="M 3425 4513 
L 3425 3897 
Q 3066 4069 2747 4153 
Q 2428 4238 2131 4238 
Q 1616 4238 1336 4038 
Q 1056 3838 1056 3469 
Q 1056 3159 1242 3001 
Q 1428 2844 1947 2747 
L 2328 2669 
Q 3034 2534 3370 2195 
Q 3706 1856 3706 1288 
Q 3706 609 3251 259 
Q 2797 -91 1919 -91 
Q 1588 -91 1214 -16 
Q 841 59 441 206 
L 441 856 
Q 825 641 1194 531 
Q 1563 422 1919 422 
Q 2459 422 2753 634 
Q 3047 847 3047 1241 
Q 3047 1584 2836 1778 
Q 2625 1972 2144 2069 
L 1759 2144 
Q 1053 2284 737 2584 
Q 422 2884 422 3419 
Q 422 4038 858 4394 
Q 1294 4750 2059 4750 
Q 2388 4750 2728 4690 
Q 3069 4631 3425 4513 
z
" id="DejaVuSans-53" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- GR -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- IT -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- LB -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- LY -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- MA -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- MT -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- SI -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- TN -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- TR -->
 
 
 
 
 
 
 
<!-- Страна -->
 
 
 
 
 
<path d="M 2194 1759 
Q 1497 1759 1228 1600 
Q 959 1441 959 1056 
Q 959 750 1161 570 
Q 1363 391 1709 391 
Q 2188 391 2477 730 
Q 2766 1069 2766 1631 
L 2766 1759 
L 2194 1759 
z
M 3341 1997 
L 3341 0 
L 2766 0 
L 2766 531 
Q 2569 213 2275 61 
Q 1981 -91 1556 -91 
Q 1019 -91 701 211 
Q 384 513 384 1019 
Q 384 1609 779 1909 
Q 1175 2209 1959 2209 
L 2766 2209 
L 2766 2266 
Q 2766 2663 2505 2880 
Q 2244 3097 1772 3097 
Q 1472 3097 1187 3025 
Q 903 2953 641 2809 
L 641 3341 
Q 956 3463 1253 3523 
Q 1550 3584 1831 3584 
Q 2591 3584 2966 3190 
Q 3341 2797 3341 1997 
z
" id="DejaVuSans-430" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{5}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{6}}$ -->
 
 
<path d="M 2113 2584 
Q 1688 2584 1439 2293 
Q 1191 2003 1191 1497 
Q 1191 994 1439 701 
Q 1688 409 2113 409 
Q 2538 409 2786 701 
Q 3034 994 3034 1497 
Q 3034 2003 2786 2293 
Q 2538 2584 2113 2584 
z
M 3366 4563 
L 3366 3988 
Q 3128 4100 2886 4159 
Q 2644 4219 2406 4219 
Q 1781 4219 1451 3797 
Q 1122 3375 1075 2522 
Q 1259 2794 1537 2939 
Q 1816 3084 2150 3084 
Q 2853 3084 3261 2657 
Q 3669 2231 3669 1497 
Q 3669 778 3244 343 
Q 2819 -91 2113 -91 
Q 1303 -91 875 529 
Q 447 1150 447 2328 
Q 447 3434 972 4092 
Q 1497 4750 2381 4750 
Q 2619 4750 2861 4703 
Q 3103 4656 3366 4563 
z
" id="DejaVuSans-36" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{7}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- Количество просм

**Ваш ответ тут:**

# Часть 3. Numpy (6 баллов), сдача в систему проверки

In [ ]:
import numpy as np

**Внимание!**

- Файлы research_functions.py и research_functions_vectorised.py с реализованными функциями необходимо сдать на тестирование в проверяющей системе.

### **Задачи 1-5 [5 баллов]**

Для каждой из задач ниже привидите 2 реализации: одна без использования numpy, а вторая полностью векторизованная ( без использования циклов/map/list comprehension ).
Реализации без использования векторизации нужно записать в файл
research_functions.py, а векторизованные — в файл
research_functions_vectorised.py (см. шаблоны).

**Внимание!**

Векторизованные реализации с использованием циклов/map/list comprehension оцениваются в 0 баллов .

```
*   Невекторизованные варианты оцениваются - по 0.3 балла
*   Векторизованные варианты - по 0.7 балла
```

- **Задача 1** : Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.
**Пример**
Input: x = [1, 2, 2, 4], y = [4, 2, 1, 2]
Output: True

- **Задача 2** : Найти максимальное
прозведение соседних элементов в массиве x, таких что хотя бы один
множитель в произведении делится на 3. Если таких произведений нет, то
вернуть -1.
**Пример**
Input: x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
Output: 12

- **Задача 3** : Дан трёхмерный
массив image, содержащий изображение, размера (height, width,
num_channels), а также вектор весов weights длины num_channels. Сложить
каналы изображения с указанными весами, и вернуть результат в виде
матрицы размера (height, width). Обратите внимание, что в изображении может быть не три канала!
**Пример**
Input: image = [[[1, 1, 1], [2, 2, 2], [3, 3, 3]], [[4, 4, 4], [5, 5, 5], [6, 6, 6]], [[7, 7, 7], [8, 8, 8], [9, 9, 9]]], weights= [1, 2, 3]
Output: [[6, 12, 18], [24, 30, 36], [42, 48, 54]]
**Комментарий**
В примере исходное изображение с тремя каналами, где по всем каналам значения равны [[1, 2, 3], [4, 5, 6], [7, 8, 9]].
После умножения первого канала на вес 1 получим значения для первого канала [[1, 2, 3], [4, 5, 6], [7, 8, 9]].
После умножения второго канала на вес 2 получим значения для второго канала [[2, 4, 6], [8, 10, 12], [14, 16, 18]].
После умножения третьего канала на вес 3 получим значения для третьего канала [[3, 6, 9], [12, 15, 18], [21, 24, 27]].
Если сложить полученное изображение поканально получим [[6, 12, 18], [24, 30, 36], [42, 48, 54]].

- **Задача 4** : Найти скалярное
произведение между векторами x и y, заданными в формате RLE (Run-length
encoding, кодирование длин серий). Каждый закодированный вектор
представлен двумерным массивом, каждая строка которого имеет размерность
2, где первое число - элемент, а второе число - сколько раз элемент
нужно повторить. Например, [[1,2], [2, 3], [3, 1]] соответствует веткору
[1, 1, 2, 2, 2, 3]. В случае несовпадения длин исходных векторов
вернуть -1.
**Пример**
Input: x = [[1, 2], [2, 3], [3, 1]], y = [[1, 1], [0, 5]].
Output: 1
- **Задача 5** : Даны две выборки
объектов - X и Y. Вычислить матрицу косинусных расстояний между
объектами X и Y. Пусть X - матрица n объектов размера (n, d), Y -
матрица из m объектов размера (m, d), тогда матрица косинусных
расстояний M - матрица размера (n, m), где m_ij - косинус угла между i-м
объектом из матрицы Х (i-ая строка) и j-м объектом из матрицы Y (j-ая
строка матрицы Y). Если хотя бы один из векторов равен 0, то косинусное
расстояние определим равным 1.
**Пример**
Input: X = [[0, 0, 0], [1, 0, 0]], Y = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
Output: [[1, 1, 1], [1, 0, 0]]

**Примечание** Можно считать, что все входные данные непустые и коректные.

**Внимание!**

- Следующее задание оценивается по системе кросс-рецензирования .

### **Задача 6 [1 балл]**

- **(0.5 балла)** Для каждой задачи
сравните скорость работы невекторизованной и векторизованной реализации.
Для этого постройте графики времени работы векторизованной и
невекторизованной реализации в зависимости от размера данных ( line plot ). Код генерации данных разных размеров и функция для замера времени работы функций уже написаны.
- **(0.5 баллa)** Сделайте выводы о времени работы функций.

**Внимание!** В текущую директорию
необходимо загрузить файлы research_functions.py и
research_functions_vectorised.py с реализованными функциями. В Google
Colab это можно сделать, нажав на значок "папки" слева, дальше на
"загрузить" (значок листа со стрелкой вверх). Загружать нужно в папку
"sample_data" (она у вас откроется по умолчанию)

In [ ]:
!ls
import time
import numpy as np

import research_functions_vectorised
import research_functions

requirements_2024_25_for_colab_small.txt  sample_data
research_functions.py			  youtube_trending_mediterranean.csv
research_functions_vectorised.py


In [ ]:
# генерация данных

are_multisets_equal_data = []
for n in [10, 100, 1000, 10000]:
    x = np.random.randint(0, 2, n)
    y = np.random.randint(0, 2, n)
    are_multisets_equal_data += [(x, y)]

max_prod_mod_3_data = []
for n in [10, 100, 1000, 10000]:
    x = np.random.randint(1, 100, n)
    max_prod_mod_3_data += [x]

convert_image_data = []
for n in [10, 100, 1000]:
    image = np.random.random((n, n, 10))
    weights = np.random.random(10)
    convert_image_data += [(image, weights)]

rle_scalar_data = []
for n in [10, 100, 1000, 10000]:
    x = np.random.randint(1, 3, (n, 2))
    y = np.random.randint(1, 3, (n, 2))
    shapes = np.random.randint(1, 3, n)
    x[:, 1] = shapes
    y[:, 1] = shapes

    rle_scalar_data += [(x, y)]

cosine_distance_data = []
for n in [10, 100]:
    X = np.random.random((n, n))
    Y = np.random.random((n, n))
    cosine_distance_data += [(X, Y)]

In [ ]:
# функция для подсчета времени работы

def count_time(func, *data):
    tick = time.perf_counter()
    func(*data)
    return time.perf_counter() - tick

In [ ]:
# получим время работы функций

# max_prod_mod_3

max_prod_mod_3_time = []
max_prod_mod_3_vectorised_time = []

for x in max_prod_mod_3_data:
  max_prod_mod_3_time += [count_time(research_functions.max_prod_mod_3, x.tolist())]
  max_prod_mod_3_vectorised_time += [count_time(research_functions_vectorised.max_prod_mod_3, x)]

sizes_max_prod_mod_3 = [10, 100, 1000, 10000]

# are_multisets_equal
are_multisets_equal_time = []
are_multisets_equal_vectorised_time = []

for x, y in are_multisets_equal_data:
    are_multisets_equal_time += [count_time(research_functions.are_multisets_equal, x.tolist(), y.tolist())]
    are_multisets_equal_vectorised_time += [count_time(research_functions_vectorised.are_multisets_equal, x, y)]

sizes_are_multisets_equal = [10, 100, 1000, 10000]

# convert_image
convert_image_time = []
convert_image_vectorised_time = []

for image, weights in convert_image_data:
    convert_image_time += [count_time(research_functions.convert_image, image.tolist(), weights.tolist())]
    convert_image_vectorised_time += [count_time(research_functions_vectorised.convert_image, image, weights)]

sizes_convert_image = [10, 100, 1000]

# rle_scalar
rle_scalar_time = []
rle_scalar_vectorised_time = []

for x, y in rle_scalar_data:
    rle_scalar_time += [count_time(research_functions.rle_scalar, x.tolist(), y.tolist())]
    rle_scalar_vectorised_time += [count_time(research_functions_vectorised.rle_scalar, x, y)]

sizes_rle_scalar = [10, 100, 1000, 10000]

# cosine_distance
cosine_distance_time = []
cosine_distance_vectorised_time = []

for X, Y in cosine_distance_data:
    cosine_distance_time += [count_time(research_functions.cosine_distance, X.tolist(), Y.tolist())]
    cosine_distance_vectorised_time += [count_time(research_functions_vectorised.cosine_distance, X, Y)]

sizes_cosine_distance = [10, 100]

In [ ]:
# ВАШИ ГРАФИКИ ЗДЕСЬ
import matplotlib.pyplot as plt

def plot_in_subplot(ax, sizes, non_vectorised_time, vectorised_time, title):
    ax.plot(sizes, non_vectorised_time, label='Невекторизованная реализация', marker='o', color='red')
    ax.plot(sizes, vectorised_time, label='Векторизованная реализация', marker='o', color='green')
    ax.set_xlabel('Объем данных')
    ax.set_ylabel('Время выполнения (с)')
    ax.set_xscale('log')
    ax.set_title(title)
    ax.legend()
    ax.grid(True)

fig, axs = plt.subplots(3, 2, figsize=(10, 10))

plot_in_subplot(axs[0, 0], sizes_max_prod_mod_3, max_prod_mod_3_time, max_prod_mod_3_vectorised_time, 'Max Prod Mod 3')
plot_in_subplot(axs[0, 1], sizes_are_multisets_equal, are_multisets_equal_time, are_multisets_equal_vectorised_time, 'Are Multisets Equal')
plot_in_subplot(axs[1, 0], sizes_convert_image, convert_image_time, convert_image_vectorised_time, 'Convert Image')
plot_in_subplot(axs[1, 1], sizes_rle_scalar, rle_scalar_time, rle_scalar_vectorised_time, 'RLE Scalar')
plot_in_subplot(axs[2, 0], sizes_cosine_distance, cosine_distance_time, cosine_distance_vectorised_time, 'Cosine Distance')

fig.delaxes(axs[2, 1])
plt.tight_layout()

?xml version="1.0" encoding="utf-8" standalone="no"?
 
 
 
 
 
 2024-10-11T17:42:07.160302 
 image/svg+xml 
 
 
 Matplotlib v3.7.1, https://matplotlib.org/ 
 
 
 
 
 
 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{1}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{2}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{3}}$ -->
 
 
<path d="M 2597 2516 
Q 3050 2419 3304 2112 
Q 3559 1806 3559 1356 
Q 3559 666 3084 287 
Q 2609 -91 1734 -91 
Q 1441 -91 1130 -33 
Q 819 25 488 141 
L 488 750 
Q 750 597 1062 519 
Q 1375 441 1716 441 
Q 2309 441 2620 675 
Q 2931 909 2931 1356 
Q 2931 1769 2642 2001 
Q 2353 2234 1838 2234 
L 1294 2234 
L 1294 2753 
L 1863 2753 
Q 2328 2753 2575 2939 
Q 2822 3125 2822 3475 
Q 2822 3834 2567 4026 
Q 2313 4219 1838 4219 
Q 1578 4219 1281 4162 
Q 984 4106 628 3988 
L 628 4550 
Q 988 4650 1302 4700 
Q 1616 4750 1894 4750 
Q 2613 4750 3031 4423 
Q 3450 4097 3450 3541 
Q 3450 3153 3228 2886 
Q 3006 2619 2597 2516 
z
" id="DejaVuSans-33" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{4}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- Объем данных -->
 
 
 
<path d="M 1991 3584 
Q 2741 3584 3169 3096 
Q 3597 2609 3597 1747 
Q 3597 888 3169 398 
Q 2741 -91 1991 -91 
Q 1238 -91 822 384 
Q 406 859 384 1747 
L 366 2463 
Q 350 2834 350 2963 
Q 350 3463 531 3831 
Q 816 4409 1308 4659 
Q 1800 4909 2625 4916 
Q 3006 4919 3175 4972 
L 3334 4516 
Q 3181 4453 3041 4450 
L 2259 4397 
Q 1866 4369 1628 4231 
Q 1006 3866 947 3388 
L 925 3213 
Q 1331 3584 1991 3584 
z
M 1991 3097 
Q 1528 3097 1259 2736 
Q 991 2375 991 1747 
Q 991 1119 1258 758 
Q 1525 397 1991 397 
Q 2450 397 2719 759 
Q 2988 1122 2988 1747 
Q 2988 2369 2719 2733 
Q 2450 3097 1991 3097 
z
" id="DejaVuSans-431" transform="scale(0.015625)"> 
 
 
 
 
 
<path d="M 2194 1759 
Q 1497 1759 1228 1600 
Q 959 1441 959 1056 
Q 959 750 1161 570 
Q 1363 391 1709 391 
Q 2188 391 2477 730 
Q 2766 1069 2766 1631 
L 2766 1759 
L 2194 1759 
z
M 3341 1997 
L 3341 0 
L 2766 0 
L 2766 531 
Q 2569 213 2275 61 
Q 1981 -91 1556 -91 
Q 1019 -91 701 211 
Q 384 513 384 1019 
Q 384 1609 779 1909 
Q 1175 2209 1959 2209 
L 2766 2209 
L 2766 2266 
Q 2766 2663 2505 2880 
Q 2244 3097 1772 3097 
Q 1472 3097 1187 3025 
Q 903 2953 641 2809 
L 641 3341 
Q 956 3463 1253 3523 
Q 1550 3584 1831 3584 
Q 2591 3584 2966 3190 
Q 3341 2797 3341 1997 
z
" id="DejaVuSans-430" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0000 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0025 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0050 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0075 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0100 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0125 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0.0150 -->
 
 
 
 
 
 
 
 
 
 
 
<!-- Время выполнения (с) -->
 
 
<path d="M 1259 2228 
L 1259 519 
L 2272 519 
Q 2781 519 3026 730 
Q 3272 941 3272 1375 
Q 3272 1813 3026 2020 
Q 2781 2228 2272 2228 
L 1259 2228 
z
M 1259 4147 
L 1259 2741 
L 2194 2741 
Q 2656 2741 2882 2914 
Q 3109 3088 3109 3444 
Q 3109 3797 2882 3972 
Q 2656 4147 2194 4147 
L 1259 4147 
z
M 628 4666 
L 2241 4666 
Q 2963 4666 3353 4366 
Q 3744 4066 3744 3513 
Q 3744 3084 3544 2831 
Q 3344 2578 2956 2516 
Q 3422 2416 3680 2098 
Q 3938 1781 3938 1306 
Q 3938 681 3513 340 
Q 3088 0 2303 0 
L 628 0 
L 628 4666 
z
" id="DejaVuSans-412" transform="scale(0.015625)"> 
 
 
<path d="M 1156 1613 
L 1156 459 
L 1975 459 
Q 2369 459 2575 607 
Q 2781 756 2781 1038 
Q 2781 1319 25

**Ваш ответ тут:**